### DISCUSSION TITLE
Approach for Factory Generation

### DISCUSSION THREAD

**Discussion by cball** [echobind/prisma-factory#1] [Ideas]
Throwing a few ideas out. The goal is a way to setup factories in a project that are automatically fully typed based on the Prisma types.

## The createFactory definition

This is more manual, but it's also explicit and clear. `prisma-factory` would expose a `createFactory` function that returns an object with `build` and `create` functions.

We can also return hooks, like `beforeCreate` that allows us to do things like hash a password.

This is our easiest path, and one Echobind has already have implemented in projects with Prisma today.

---

## Automatically Generate via a prisma generator

```
generator factories {
  provider = "prisma-factories"
}
```

This would require no user-defined factories unless they want to override the attribute generators. This could make for a nice default, but users might not be clear where things come from.

Prisma docs on generators: https://github.com/prisma/specs/tree/master/generators

---

## Use Field annotation to map  functions

```
model User {
  id        Int       @default(autoincrement()) @id
  createdAt DateTime  @default(now())
  email     String    @unique @factory(chance.email)
  name      String?   @factory(chance.company)
}
```

This would let the user override functions, but wouldn't handle some of the advanced cases we're considering (like `beforeCreate` hooks)

---

**Give it a 🤔 and comment!**

**Comment by cball** [comment]
Another thought from a recent Echobind team discussion:

Could we use the prisma generator approach to automatically generate typesafe functions? For example, given the schema above, we would use a prisma generator to build a `createUserFactory` function. The user could then use it as follows:

```ts
import { createUserFactory } from 'prisma-factories'

const chance = new Chance();

const DEFAULT_ATTRIBUTES = {
  id: chance.guid(),
  slug: chance.word(),
};

export const UserFactory = createUserFactory(DEFAULT_ATTRIBUTES);
```

This feels like the best of both worlds and doesn't require the user to pass in the input type (since we can derive it)!

**Comment by Jolg42** [comment]
There is a list of community generators here for info
https://www.prisma.io/docs/concepts/components/prisma-schema/generators

**Reply by cball** [reply]
perfect. Thanks @Jolg42.

**Comment by Jolg42** [comment]
About generator specs, the `specs` repository is archived & generally outdated so I can recommend this instead https://www.notion.so/prismaio/Prisma-Generators-a2cdf262207a4e9dbcd0e362dfac8dc0

**Comment by toyamarinyon** [comment]
I created factory generator based on this idea.
https://github.com/toyamarinyon/prisma-factory-generator

I made it in two days, so it's still lacking some features and documentation, but if anyone is interested in using it, I'd be happy to get feedback!

**Reply by younes0** [reply]
Awesome.
You could also check for inspiration: https://laravel.com/docs/8.x/database-testing#defining-model-factories
